File: src/han/model/network/Network.java
package han.model.network;

import han.model.network.nodes.HiddenNode;
import han.model.network.nodes.InputNode;
import han.model.network.nodes.OutputNode;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by han on 16-11-16.
 * Network class containing nodes of the network
 */
public class Network {

    /**
     * Lists to store the nodes that make up the network
     */
    private List<Node> inputNodeGroup = new ArrayList<>();
    private List<List<Node>> hiddenNodeGroupList = new ArrayList<>();
    private List<Node> outputNodeGroup = new ArrayList<>();
    private List<Edge> edgeList = new ArrayList<>();

    /**
     * Constructor for the network when provided with all the variable amount of nodes per node group
     * @param amountOfInputNodes Amount of input nodes in input node group
     * @param amountOfHiddenNodeGroups Amount of hidden node group
     * @param amountOfHiddenNodes Amount of hidden Nodes per hidden node group
     * @param amountOfOutputNodes Amount of output Nodes per output node group
     */
    public Network(int amountOfInputNodes, int amountOfHiddenNodeGroups, int amountOfHiddenNodes,
                   int amountOfOutputNodes) {
        createNodes(amountOfInputNodes, amountOfHiddenNodeGroups, amountOfHiddenNodes, amountOfOutputNodes);
        try {
            createEdges(inputNodeGroup, hiddenNodeGroupList.get(0));
        } catch (IndexOutOfBoundsException e) {
            System.out.println("There should at least be one hidden node group");
        }
        for (int i = 1; i < hiddenNodeGroupList.size(); i++) {
            createEdges(hiddenNodeGroupList.get(i - 1), hiddenNodeGroupList.get(i));
        }
        createEdges(hiddenNodeGroupList.get(hiddenNodeGroupList.size() - 1), outputNodeGroup);
    }

    /**
     * Adds the nodes to the network
     * @param amountOfInputNodes Amount of input nodes in input node group
     * @param amountOfHiddenNodeGroups Amount of hidden node group
     * @param amountOfHiddenNodes Amount of hidden Nodes per hidden node group
     * @param amountOfOutputNodes Amount of output Nodes per output node group
     */
    private void createNodes(int amountOfInputNodes, int amountOfHiddenNodeGroups, int amountOfHiddenNodes,
                             int amountOfOutputNodes) {
        for (int i = 0; i < amountOfInputNodes; i++) {
            inputNodeGroup.add(new InputNode(0, i));
        }
        for (int i = 0; i < amountOfOutputNodes; i++) {
            outputNodeGroup.add(new OutputNode(amountOfHiddenNodeGroups + 1, i));
        }
        for (int i = 0; i < amountOfHiddenNodeGroups; i++) {
            List<Node> hiddenNodeGroup = new ArrayList<>();
            for (int j = 0; j < amountOfHiddenNodes; j++) {
                hiddenNodeGroup.add(new HiddenNode(i + 1, j));
            }
            hiddenNodeGroupList.add(hiddenNodeGroup);
        }
    }

    private void createEdges(List<Node> sourceNodeList, List<Node> destinationNodeList) {
        for (Node sourceNode : sourceNodeList) {
            for (Node destinationNode : destinationNodeList) {
                Edge edge = new Edge(sourceNode, destinationNode);
                sourceNode.addEdge(edge);
                edgeList.add(edge);
            }
        }
    }

    public List<Node> getInputNodeGroup() {
        return inputNodeGroup;
    }

    public List<List<Node>> getHiddenNodeGroupList() {
        return hiddenNodeGroupList;
    }

    public List<Node> getOutputNodeGroup() {
        return outputNodeGroup;
    }

    public List<Edge> getEdgeList() {
        return edgeList;
    }

}

File: python/src/han/model/network/network.py
from han.exception import IllegalValueException
from han.model.network.edge import Edge
from han.model.network.node import Node


class Network:
    """Class that represents the network model."""

    def __init__(self, input_nodes: 'int', hidden_node_groups: 'int', hidden_nodes: 'int', output_nodes: 'int'):
        """Initialization funciton for the Network class. input_nodes is the amount of input nodes in the network,
        hidden_node_groups is the amount of hidden node groups, hidden_nodes is the amount of hidden nodes that are
        in each group, and output_nodes is the amount of output nodes in the network.
        """

        # Make sure legal values are used.
        if input_nodes < 0 or hidden_node_groups < 1 or hidden_nodes < 1 or output_nodes < 1:
            raise IllegalValueException('All arguments need to be greater than 0 when constructing the network!')

        # Initialize lists.
        self.input_nodes = []
        self.hidden_node_groups = []
        self.output_nodes = []
        self.edges = []

        # Fill all lists with nodes.

        # Input nodes.
        for i in range(0, input_nodes):
            self.input_nodes.append(Node())

        # Hidden nodes.
        for i in range(0, hidden_node_groups):
            temp_hidden_node_group = []
            for j in range(0, hidden_nodes):
                temp_hidden_node_group.append(Node())
            self.hidden_node_groups.append(temp_hidden_node_group)

        # Output nodes.
        for i in range(0, output_nodes):
            self.output_nodes.append(Node())

        # Create all edges between nodes.

        # Create edges from input nodes to the first group of hidden nodes.
        for input_node in self.input_nodes:
            for hidden_node in self.hidden_node_groups[0]:
                edge = Edge(input_node, hidden_node, 1)
                self.edges.append(edge)
                input_node.add_output_edge(edge)

        # Create edges from hidden nodes.
        for i in range(0, len(self.hidden_node_groups)):
            hidden_node_group = hidden_node_groups[i]

            # Edges between hidden node groups.
            if i < len(self.hidden_node_groups) - 1:
                hidden_node_target_group = hidden_node_groups[i + 1]
                for hidden_node in hidden_node_group:
                    for hidden_target_node in hidden_node_target_group:
                        edge = Edge(hidden_node, hidden_target_node, 1)
                        self.edges.append(edge)
                        hidden_node.add_output_edge(edge)

            # Edges from the last hidden node group to the output node.
            elif i == len(self.hidden_node_groups) - 1:
                for hidden_node in hidden_node_group:
                    for output_node in self.output_nodes:
                        edge = Edge(hidden_node, output_node, 1)
                        self.edges.append(edge)
                        hidden_node.add_output_edge(edge)


